learn/testing-ballerina-code/writing-tests.md

```diff
@@ -0,0 +1,348 @@
+
+Asserts that the expression is false with an optional message.
+
+***Example:***
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertFalse() {
+    boolean value = false;
+    test:assertFalse(value, msg = "AssertFalse failed");
+}
+```
+
+### **assertEquals(Any actual, Any expected, string message)**
+
+Asserts that the actual is equal to the expected with an optional message.
+
+***Example:***
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertIntEquals() {
+
+    int answer = 0;
+    int a = 5;
+    int b = 3;
+    answer = intAdd(a, b);
+    test:assertEquals(answer, 8, msg = "IntAdd function failed");
+}
+
+function intAdd(int a, int b) returns (int) {
+    return (a + b);
+}
+```
+
+### **assertNotEquals(Any actual, Any expected, string message)**
+
+Asserts that the actual is not equal to the expected with an optional message.
+
+***Example:***
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function testAssertIntEquals() {
+
+    int answer = 0;
+    int a = 5;
+    int b = 3;
+    answer = intAdd(a, b);
+    test:assertNotEquals(answer, 8, msg = "Matches");
+}
+
+function intAdd(int a, int b) returns (int) {
+    return (a + b);
+}
+```
+
+### **assertFail(string message)**
+
+Fails the test. This is useful to fail a test based on a check for a condition while it is in execution.
+
+***Example:***
+
+```ballerina
+import ballerina/test;
+
+@test:Config {}
+function foo() {
+    error? e = trap bar(); // Expecting `bar()` to panic
+    if (e is error) {
+        test:assertEquals(e.reason(), "Invalid Operation", msg = "Invalid error reason"); // Some other assertions
+    } else {
+        test:assertFail(msg = "Expected an error");
+    }
+}
+```
+
+## Setup and Teardown
+
+The following test annotations can be used for setup and teardown instructions. These annotations enable executing instructions in different levels.
+
+### @test:BeforeSuite {}
+
+The function specified following the annotation will be run once before any of the tests in the test suite is run. This can be used for initializing the test suite level aspects.
+
+***Example:***
+
+```ballerina
+import ballerina/io;
+import ballerina/test;
+
+// The `BeforeSuite` function is executed before running all the test functions in this module. 
+@test:BeforeSuite
+function beforeFunc() {
+    io:println("I'm the before suite function!");
+}
+
+// Test function
+@test:Config {}
+function testFunction1() {
+    io:println("I'm in test function 1!");
+    test:assertTrue(true, msg = "Failed");
+}
+
+// Test function
+@test:Config {}
+function testFunction2() {
+    io:println("I'm in test function 2!");
+    test:assertTrue(true, msg = "Failed");
+}
+```
+
+### @test:BeforeEach {}
+
+The function specified following the annotation will be run before every test within the test suite is run. This can be used for repeatedly initializing test-level aspects before every test function.
```

Comment:
```suggestion
The function specified after the annotation will be run before every test when the test suite is run. This can be used for initializing test-level aspects repeatedly before every test function.
```